### ISSUE TITLE
generated files have no layout

### ISSUE THREAD

**Issue by oupala** [superwolff/metalsmith-layouts#26]
While 1.1.0 works well for me, updating to version 1.2.0 does not work : all html generated file contains only the content file (from markdown) but not layout at all.

Could you lead me to some point where I can debug and find where it comes from ?

First thing, I notice that, in debug mode, version 1.1.0 the build output the following :

```
contents: <Buffer 3c 21 44 4f 43 54 59 50 45 20 68 74 6d 6c 3e 0a 3c 68 74 6d 6c 20 63 6c 61 73 73 3d 22 77 66 2d 70 74 73 61 6e 73 2d 6e 34 2d 61 63 74 69 76 65 20 77 66 ...>,
```

while in version 1.2.0, I get the following output :

```
contents: '<h1 id="first-header"><a class="heading-anchor" href="#first-header"><span></span></a>[...]
```

**Comment by ismay** [comment]
The only big change is that binary files are ignored in 1.2.0, and not in 1.1.0. A reduced test case would help in solving this.

**Comment by oupala** [comment]
I tried to update your plugin to version 1.2.1 but it was not working either.

I then tried some reduced test case, as you advised to me, and I finally manage to find the incompatibility.

I then tried some reduced test case, as you have advised me, and I managed to find the source of the problem. When metalsmith-headings-identifier is used before metalsmith-layouts, the layout is not applied (I don't know why). If I change my build file and call metalsmith-headings-identifier after metalsmith-layouts, everything seems to be ok.

Could this be caused by metalsmith-headings-identifier which is maybe outputing a string instead of a buffer ?

**Comment by ismay** [comment]
Yeah, metalsmith parses files to buffers, from the [readme](https://github.com/segmentio/metalsmith#how-does-it-work): 

> a simple file like...

>```
>---
>title: A Catchy Title
>date: 2014-12-01
>---
>
>An informative article.
>```
> ...would be parsed into...
>```
>{
>  'path/to/my-file.md': {
>    title: 'A Catchy Title',
>    date: new Date('2014-12-01'),
>    contents: new Buffer('An informative article.')
>  }
>}
>```

So a metalsmith plugin like `metalsmith-layouts` expects `file.contents` to [be a buffer](https://github.com/superwolff/metalsmith-layouts/blob/master/lib/index.js#L63) and returns it [as a buffer](https://github.com/superwolff/metalsmith-layouts/blob/master/lib/index.js#L79). And it looks like there's already an issue on the `metalsmith-headings-identifier` repo about this exact problem: https://github.com/majodev/metalsmith-headings-identifier/issues/4

**Comment by ismay** [comment]
I'll close this, but feel free to reopen if this hasn't resolved the issue!

**Comment by oupala** [comment]
Ok for closing this issue, as the bug seems to come from metalsmith-headings-identifier.

But that doesn't explain why I had no problem with metalsmith-layouts 1.1.0 and the exact same configuration (and then same version of metalsmith-headings-identifier).

**Comment by ismay** [comment]
> But that doesn't explain why I had no problem with metalsmith-layouts 1.1.0 and the exact same configuration (and then same version of metalsmith-headings-identifier).

That's true. I'd be happy to take a look at it if you could point me to a repo or branch where you've reproduced the problem.

**Comment by oupala** [comment]
I can't easily publish my project as it is a private one. But I'll try to reproduce the problem in a minimal example.

This is only for posterity, of for curiosity that might lead to discover another hidden bug !

**Comment by ismay** [comment]
> But I'll try to reproduce the problem in a minimal example.

Cool, that would be perfect.

> This is only for posterity, of for curiosity that might lead to discover another hidden bug !

Exactly.

**Comment by oupala** [comment]
I spent some time trying to reproduce the bug, but I had a very hard time.

Because of some other migration with major upgrade (metalsmith, nodejs...), I got new errors while trying to revert metalsmith-layout and metalsmith-headings-identifier to some older version.

I'm sorry to say that I give up trying to reproduce the bug.

Nevertheless, the balance is good as we found a bug in another plugin, and that's the most important thing to retain.
